spark3/src/main/java/org/apache/iceberg/spark/source/SparkMicroBatchStream.java

```diff
@@ -0,0 +1,306 @@
+
+    return initialOffset;
+  }
+
+  @Override
+  public Offset deserializeOffset(String json) {
+    return StreamingOffset.fromJson(json);
+  }
+
+  @Override
+  public void commit(Offset end) {
+  }
+
+  @Override
+  public void stop() {
+  }
+
+  private List<FileScanTask> getFileScanTasks(StreamingOffset startOffset, StreamingOffset endOffset) {
+    List<FileScanTask> fileScanTasks = new ArrayList<>();
+    MicroBatch latestMicroBatch = null;
+    do {
+      final StreamingOffset currentOffset =
+          latestMicroBatch != null && latestMicroBatch.lastIndexOfSnapshot() ?
+          getNextAvailableSnapshot(latestMicroBatch.snapshotId()) :
+          startOffset;
+
+      latestMicroBatch = MicroBatches.from(table.snapshot(currentOffset.snapshotId()), table.io())
+          .caseSensitive(caseSensitive)
+          .specsById(table.specs())
+          .generate(currentOffset.position(), Long.MAX_VALUE, currentOffset.shouldScanAllFiles());
+
+      fileScanTasks.addAll(latestMicroBatch.tasks());
+    } while (latestMicroBatch.snapshotId() != endOffset.snapshotId());
+
+    return fileScanTasks;
+  }
+
+  private StreamingOffset getNextAvailableSnapshot(long snapshotId) {
+    Snapshot previousSnapshot = table.snapshot(snapshotId);
+    Snapshot pointer = table.currentSnapshot();
+    while (pointer != null && previousSnapshot.snapshotId() != pointer.parentId()) {
+      Preconditions.checkState(pointer.operation().equals(DataOperations.APPEND),
+          "Encountered Snapshot DataOperation other than APPEND.");
+
+      pointer = table.snapshot(pointer.parentId());
+    }
+
+    Preconditions.checkState(pointer != null,
+        "snapshot on which the stream operated has been garbage collected.");
+
+    return new StreamingOffset(pointer.snapshotId(), 0L, false);
+  }
+
+  interface OffsetLog {
+    static OffsetLog getInstance(SparkSession spark, String checkpointLocation) {
+      return new OffsetLogImpl(spark, checkpointLocation);
+    }
+
+    void addInitialOffset(StreamingOffset offset);
+
+    boolean isOffsetLogInitialized();
+
+    StreamingOffset getLatest();
+  }
+
+  private static class OffsetLogImpl implements OffsetLog {
+    private final IcebergSourceOffsetLog offsetSeqLog;
+
+    OffsetLogImpl(SparkSession spark, String checkpointLocation) {
+      this.offsetSeqLog = checkpointLocation != null ?
+          new IcebergSourceOffsetLog(spark, checkpointLocation) :
+          null;
+    }
+
+    @Override
+    public void addInitialOffset(StreamingOffset offset) {
+      this.offsetSeqLog.add(0, offset);
+    }
+
+    @Override
+    public boolean isOffsetLogInitialized() {
+      return offsetSeqLog != null &&
+          offsetSeqLog.getLatest() != null &&
+          offsetSeqLog.getLatest().isDefined();
+    }
+
+    @Override
+    public StreamingOffset getLatest() {
+      return offsetSeqLog.getLatest().get()._2;
+    }
+  }
+
+  private static class IcebergSourceOffsetLog extends HDFSMetadataLog<StreamingOffset> {
```

Comment:
This may cause us issues in the future, HDFSMetadataLog is a little internal but it extends Logging which is very internal and has burned me in the past. That said I think this would be good re-use .... Anyone else have strong feelings on this?

Reply:
One other issue, we can't support S3 based paths with this

Reply:
> This may cause us issues in the future, HDFSMetadataLog is a little internal but it extends Logging which is very internal and has burned me in the past. That said I think this would be good re-use .... Anyone else have strong feelings on this?

This came as a recommendation from one of the spark experts tagged by @rdblue & a sample from Spark codebase : https://github.com/apache/spark/blob/0494dc90af48ce7da0625485a4dc6917a244d580/sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/sources/RateStreamMicroBatchStream.scala#L64

can you pl. help me understand - what is the precise concern about using `HDFSMetadataLog`. Is it that - it is internal & prone to change?

Reply:
> One other issue, we can't support S3 based paths with this
> 
> ```
> Note: HDFSMetadataLog doesn't support S3-like file systems as they don't guarantee listing files in a directory always shows the latest files.
> ```

This checks out in our internal tests with S3 based checkpoint.

+ AWS S3 documentation now says - S3 doesn't have that file listing problem anymore: https://aws.amazon.com/s3/consistency/

```
S3 also provides strong consistency for list operations, so after a write, you can immediately perform a listing of the objects in a bucket with any changes reflected. 
```

Reply:
here's a screenshot from my testing where I resumed streaming read from an s3 based checkpoint
![image](https://user-images.githubusercontent.com/11813428/121218679-ce032080-c837-11eb-864e-c00d9a3da003.png)

Reply:
A single test is probably not sufficient :) The issue being that the list operation in S3 was eventually consistent at least before. I know they fixed some of those issues but I think they were just around reading your writes, not around the list operation ... maybe someone with more S3 knowledge than me can chime in

Reply:
Oh i see in your link they say list is fixed, then we should be fine. My concern about using this class is that it's internal to Spark, so we would be subject to breakage if say we build against Spark 3.0.1 and they make a change in Spark 3.0.2

Reply:
@aokolnychyi / @rdblue / @jackye1995 / @holdenk - this is one place where @RussellSpitzer recommended to collect consensus - in our offline chat.
**Here's the quick summary of this thread:**

1. In the connector - we are needing to checkpoint the Initial Offset; 
2. current approach in this PR uses `HDFSMetadataLog` - to store this; primary inspired from [kafka MicroBatchStream source implementation](https://github.com/apache/spark/blob/master/external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchStream.scala).
3. however, Russell pointed one very critical point - that:
    - one of the goals of iceberg is to generate single build of Iceberg for a given Spark major version.
    - But, given that `HDFSMetadataLog` - class belongs to spark - execution package - will spark potentially break this contract across minor versions!?

If you folks see this as a potential threat - pl. also recommend options to be able to persist just the `InitialOffset`.

Reply:
Yes S3 is now strongly consistent in listing and read-after-write, so that warning message is out of date.

For the point that Russell pointed out, I remember we had a discussion in the last sync meeting about if we would like to have 1 runtime for 1 Spark minor release. I don't recall we arrived at a conclusion on that. If we decide to go with that approach, then this method is fine.

If we do not go with that approach, then it is probably gonna be a trouble in the future. Because S3 is not a concern anymore, we can directly write the file through HDFS file system API, or even better the Iceberg FileIO API. But that will be a bit of duplication with the Spark codebase and some more effort in this PR.

Reply:
removed the dependency on `HDFSMetadataLog` & used `FileIO`. @jackye1995 & @RussellSpitzer - pl. take a peek & lemme know if I got that right.